File: backend/app/schemas/glossary.py
"""Pydantic schemas for GlossaryEntry endpoints."""

from __future__ import annotations

import json
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class GlossaryEntryRead(BaseModel):
    """A single glossary row returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: str
    english_term: str
    persian_term: str
    english_definition: str
    persian_definition: str
    source: str
    score: float
    frequency: int
    context: str
    is_edited: bool
    # --- AI Mode extras ---
    pos: str | None = None
    category: str | None = None
    persian_alternatives: list[str] | None = None
    translator_note: str | None = None
    persian_transliteration: str | None = None
    # --- Timestamps ---
    created_at: datetime
    updated_at: datetime

    @field_validator("persian_alternatives", mode="before")
    @classmethod
    def _parse_persian_alternatives(cls, v):
        """Accept either a JSON string (from ORM) or a list."""
        if v is None:
            return None
        if isinstance(v, list):
            return v
        if isinstance(v, str):
            if not v.strip():
                return None
            try:
                parsed = json.loads(v)
                if isinstance(parsed, list):
                    return [str(x) for x in parsed if x]
                return None
            except json.JSONDecodeError:
                return None
        return None


class GlossaryEntryUpdate(BaseModel):
    """Body of PATCH /api/glossary/{id} — manual edits."""

    persian_term: str | None = Field(default=None, max_length=255)
    persian_definition: str | None = None
    english_definition: str | None = None
    translator_note: str | None = None
    is_edited: bool | None = None


class GlossaryResponse(BaseModel):
    """Response for GET /api/projects/{id}/glossary."""

    project_id: str
    entries: list[GlossaryEntryRead]
    total: int

File: frontend/src/types/glossary.ts
export interface GlossaryEntryRead {
  id: number
  project_id: string
  english_term: string
  persian_term: string
  english_definition: string
  persian_definition: string
  source: string
  score: number
  frequency: number
  context: string
  is_edited: boolean
  // --- AI Mode extras (nullable) ---
  pos: string | null
  category: string | null
  persian_alternatives: string[] | null
  translator_note: string | null
  persian_transliteration: string | null
  // --- Timestamps ---
  created_at: string
  updated_at: string
}

export interface GlossaryEntryUpdate {
  persian_term?: string
  persian_definition?: string
  english_definition?: string
  translator_note?: string
  is_edited?: boolean
}

export interface GlossaryResponse {
  project_id: string
  entries: GlossaryEntryRead[]
  total: number
}

export type ExportFormat = 'csv' | 'xlsx' | 'tbx'